File: Springboot_Implementation/groceryapp/src/main/java/com/example/groceryapp/service/ProductService.java
package com.example.groceryapp.service;

import com.example.groceryapp.entity.Product;
import com.example.groceryapp.repository.ProductRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class ProductService {

    @Autowired
    private ProductRepository productRepository;

    // CREATE
    public Product createProduct(Product product) {
        return productRepository.save(product);
    }

    // READ ALL
    public List<Product> getAllProducts() {
        return productRepository.findAll();
    }

    // READ BY ID
    public Optional<Product> getProductById(Integer id) {
        return productRepository.findById(id);
    }

    // UPDATE
    public Product updateProduct(Integer id, Product productDetails) {

        Product product = productRepository.findById(id).orElse(null);

        if (product != null) {
            product.setProductName(productDetails.getProductName());
            product.setProductDescription(productDetails.getProductDescription());
            product.setProductPrice(productDetails.getProductPrice());
            product.setStockQuantity(productDetails.getStockQuantity());
            product.setProductOffer(productDetails.getProductOffer());
            product.setCategory(productDetails.getCategory());
            product.setStore(productDetails.getStore());

            return productRepository.save(product);
        }

        return null;
    }

    // DELETE
    public void deleteProduct(Integer id) {
        productRepository.deleteById(id);
    }

    // SEARCH BY NAME
    public List<Product> searchByName(String name) {
        return productRepository.findByProductNameContaining(name);
    }
}

File: React_Implementation/grocery-app/frontend/src/services/productService.js
import api, { firstSuccess } from './api';

export const productService = {
  getAllProducts: async () => (await firstSuccess([
    () => api.get('/api/products'),
    () => api.get('/products'),
  ])).data,
  getProductById: async (id) => (await firstSuccess([
    () => api.get(`/api/products/${id}`),
    () => api.get(`/products/${id}`),
  ])).data,
  createProduct: async (product) => (await firstSuccess([
    () => api.post('/api/products', product),
    () => api.post('/products', product),
  ])).data,
  updateProduct: async (id, product) => (await firstSuccess([
    () => api.put(`/api/products/${id}`, product),
    () => api.put(`/products/${id}`, product),
  ])).data,
  deleteProduct: async (id) => (await firstSuccess([
    () => api.delete(`/api/products/${id}`),
    () => api.delete(`/products/${id}`),
  ])).data,
  searchByName: async (name) => {
    const encoded = encodeURIComponent(name);
    return (await firstSuccess([
      () => api.get(`/api/products/search/${encoded}`),
      () => api.get(`/api/products/search?name=${encoded}`),
      () => api.get(`/products/search/${encoded}`),
    ])).data;
  },
};
